src/attribution/onFCP.ts

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getBFCacheRestoreTime} from '../lib/bfcache.js';
+import {getLoadState} from '../lib/getLoadState.js';
+import {getNavigationEntry} from '../lib/getNavigationEntry.js';
+import {onFCP as unattributedOnFCP} from '../onFCP.js';
+import {FCPMetricWithAttribution, FCPReportCallback, FCPReportCallbackWithAttribution, ReportOpts} from '../types.js';
+
+
+const attributeFCP = (metric: FCPMetricWithAttribution): void => {
+  if (metric.entries.length) {
+    const navigationEntry = getNavigationEntry();
+
+    if (navigationEntry) {
+      const activationStart = navigationEntry.activationStart || 0;
+      const ttfb = Math.max(0, navigationEntry.responseStart - activationStart);
+
+      metric.attribution = {
+        timeToFirstByte: ttfb,
+        renderDelay: metric.value - ttfb,
```

Comment:
Is it reasonable to name this renderDelay?

Typically render delay is the time after content is available, but before it is presented.  TTFB is not the time after content is available, so this value is both the sum of some loadTime + renderDelay.

I don't know if it's possible to define loadTime for FCP, so a combined metric may make sense... does is the name renderDelay misleading?

Reply:
I struggled a lot with naming on this one.

I agree that "delay" here doesn't fit with how delay is used in other attribution properties. I was also considering `renderTime`, but `renderTime` is a spec'ed property of the LCP entry, and there it's defined as a timestamp and not a duration, so I thought that might be confusing.

I went with `renderDelay` for consistency with `elementRenderDelay` in LCP attribution, but I agree that's not an entirely apples-to-apples comparison.

Anyway, all that to say, I'm open to renaming this (and others), but I don't currently have a better name in mind, so I'm open to suggestions.

Reply:
What I meant was `render` part of the name, not `Delay` part.  I don't mind using `Delay` since I dislike using `Time` for durations.

I think renderDelay here is different than elementRenderDelay, in that FCP really also has a `resourceLoadDelay` and `resourceLoadTime` its just less well defined.  This value if lumping them all together.

Reply:
In the case of FCP, I'm not sure that we'd want to break the time down further? At least not in a way that's specific to resource load given that pages can reach FCP without loading any resources (and that's historically what we've recommended, e.g. remove render-blocking resources).

Reply:
If there is text content, then the main document is the resource.  If there is no text, then we will not hit FCP until the first image is painted, and so it may not be render-blocking but its still FCP-blocking so to speak.

Either way, whatever the resource, the specific paint that leads to FCP you can always at the very least break down to the time it took to get to the start of the paint operation, and the additional time it took to present the frame from there.  This helps find cases where rAF, style, layout, paint, raster, decode, or draw end up taking a long time.

Reply:
> You can always at the very least break down to the time it took to get to the start of the paint operation, and the additional time it took to present the frame from there.

But none of that information is exposed via any RUM APIs, correct?

I agree that someone  trying to diagnose FCP issues would look at all of this information in a DevTools trace, but from a RUM analytics perspective, I think the most useful signal is the high-level: 1) is FCP slow due to network or server slowness? Or 2) is FCP slow due to front-end rendering bottlenecks.

I think if/when we have the [render blocking status](https://chromestatus.com/feature/5166965277589504) in Resource Timing, we could add that as an additional signal, but otherwise I'm not sure how else we could break down the `renderDelay` portion generically given existing RUM APIs.

Reply:
To reiterate, I have no concerns with how you have defined this value here for FCP.  I think it's the most reasonable way to attribute today.

I'm really just expressing a soft-concern about using the term "renderDelay" to mean a slightly different thing in this context than it is elsewhere... especially since it's conceivable we'll want to split it out sometimes.

As you say, you can do this split out in the lab-- so if we think that the term "renderDelay" would be a good name for the final subset of the time in lab tooling-- even just for consistency reasons alone, I think the field value should be distinct.

(Also: It is not available to RUM via paint timing yet but I think there is some hope that it could be in the future.  And, even today if you are listening to rAF timings you can sorta compute it yourself).

Reply:
Sorry for the bikeshed.

Of those options, I think I like `responseToRender` best.

Other options:
* `firstByteToFCP` (or replace FCP with paint/render/etc)
* `loadAndRenderDelay`

Reply:
Let's go with `firstByteToFCP`. It's _very_ clear what timestan it covers, and if in the future we're able to break it down further, we could add additional sub-parts without causing too much confusion or forcing a breaking change.